### PR TITLE
Able to use "SearchResult.next()" for securities

### LINKED ISSUE #26: Implement ProfileSearchResult.next()
Implement ProfileSearchResult.next()

### LINKED ISSUE #27: Implement RoleSearchResult.next()
Implement RoleSearchResult.next()

### LINKED ISSUE #28: Implement UsersSearchResult.next()
Implement UsersSearchResult.next()

### PR DESCRIPTION
**Pull request by y-abs**
## What does this PR do ?

This PR permits to use `RoleSearchResult.next()`, `ProfileSearchResult.next()` and `UserSearchResult.next()`. 

https://github.com/kuzzleio/sdk-dart/issues/26
https://github.com/kuzzleio/sdk-dart/issues/27
https://github.com/kuzzleio/sdk-dart/issues/28

### How should this be manually tested?

Snippets

### Other changes

Function `buildNextSearchResult` now is `protected` instead of `private` to override it in its subclasses.

### Boyscout

Remove `scroll` documentation for `searchRoles` since there is no `scrollAction` for it.
Rename `abstract.dart` to `search-result.dart` since `SearchResult` class is not abstract.

### CODE REVIEW

File: doc/2/core-classes/profile/next/snippets/fromsize.dart [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    [
+      {
+        'roleId': 'default'
+      }
+    ]
+  );
+}
+
+var res = await kuzzle.security.searchProfiles(query:{
+  'roles': ['default']}, from: 1, size: 1);
```
Comment (scottinet) [inline doc/2/core-classes/profile/next/snippets/fromsize.dart]: (nitpicking) more readable line wrapping:

```suggestion
var res = await kuzzle.security.searchProfiles(
  query:{ roles': [ 'default' ] }, 
  from: 1, 
  size: 1);
```

File: doc/2/core-classes/profile/next/snippets/scroll.dart [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    [
+      {
+        'roleId': 'default'
+      }
+    ]
+  );
+}
+
+var res = await kuzzle.security.searchProfiles(query:{
+  'roles': ['default']}, scroll: '10s', size: 1);
```
Comment (scottinet) [inline doc/2/core-classes/profile/next/snippets/scroll.dart]: (nitpicking) more readable line wrapping:

```suggestion
var res = await kuzzle.security.searchProfiles(
  query:{ 'roles': [ 'default' ] }, 
  scroll: '10s', 
  size: 1);
```

File: doc/2/core-classes/role/next/index.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+## Returns
+
+Returns a `RoleSearchResult` object, or `null` if no more pages are available.
+
+## Pagination strategies
+
+Depending on the arguments given to the initial search, the `next` method will pick one of the following strategies, by decreasing order of priority.
+
+### Strategy: scroll cursor
```
Comment (scottinet) [inline doc/2/core-classes/role/next/index.md]: `security:searchRoles` does neither support scroll cursors nor sort/size searches.

File: doc/2/core-classes/role/next/snippets/fromsize.dart [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+          'getMyRights': true,
+          'logout': true
+        }
+      }
+    }
+  );
+}
+
+var res = await kuzzle.security.searchRoles(query:{
+  'controllers': ['auth']}, from: 1, size: 1);
```
Comment (scottinet) [inline doc/2/core-classes/role/next/snippets/fromsize.dart]: (nitpicking)

```suggestion
var res = await kuzzle.security.searchRoles(
  query:{ 'controllers': [ 'auth' ] }, 
  from: 1, 
  size: 1);
```

File: doc/2/core-classes/user/next/index.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
+---
```
Comment (scottinet) [inline doc/2/core-classes/user/next/index.md:1]: multiple copy/paste mishaps in this page

File: doc/2/core-classes/user/next/snippets/fromsize.dart [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  });
+}
+
+var res = await kuzzle.security.searchUsers(query:
+  {
+  'query': {
+    'term': {
+      'status': 'student'
+    }
+  }}, from: 1, size: 1);
```
Comment (scottinet) [inline doc/2/core-classes/user/next/snippets/fromsize.dart]: (nitpicking)

```suggestion
var res = await kuzzle.security.searchUsers(
  query: {
    'query': {
      'term': {
        'status': 'student'
      }
    }
  }, 
  from: 1, 
  size: 1);
```

File: doc/2/core-classes/user/next/snippets/scroll.dart [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  });
+}
+
+var res = await kuzzle.security.searchUsers(query:
+  {
+  'query': {
+    'term': {
+      'status': 'student'
+    }
+  }}, scroll: '10s', size: 1);
```
Comment (scottinet) [inline doc/2/core-classes/user/next/snippets/scroll.dart]: (nitpicking)

```suggestion
var res = await kuzzle.security.searchUsers(
  query: {
    'query': {
      'term': {
        'status': 'student'
      }
    }
  }, 
  scroll: '10s', 
  size: 1);
```

File: lib/src/search_result/search-result.dart [hunk 8]
```diff
@@ -56,18 +55,22 @@ class SearchResult {
...
     if (request.scroll != null && request.scroll.isNotEmpty) {
-      return await kuzzle
-          .query(KuzzleRequest(
+      final query = KuzzleRequest(
         controller: controller,
-        action: scrollAction,
         scrollId: response.result['scrollId'] as String,
-      ))
+      );
+      if (searchAction != 'searchRoles') {
```
Comment (scottinet) [inline lib/src/search_result/search-result.dart]: bad practice: the parent class is not supposed to know anything about classes that inherit it

this should be handled by an override in the role's child class

**Review by scottinet — CHANGES_REQUESTED** [review]

File: lib/src/search_result/roles.dart [hunk 9]
```diff
@@ -31,5 +31,113 @@ class RoleSearchResult extends SearchResult {
     return nextSearchResult;
   }
 
+  @override
+  Future<SearchResult> next() async {
+    if (fetched >= total) {
+      return null;
+    }
+
+    if (request.scroll != null && request.scroll.isNotEmpty) {
```
Comment (scottinet) [inline lib/src/search_result/roles.dart]: Pagination on roles search does neither support scrolls nor sort+size strategies.

**Review by scottinet — CHANGES_REQUESTED** [review]

File: lib/src/search_result/roles.dart (hunk 9 above)
Comment (y-abs) [reply lib/src/search_result/roles.dart]: -_- my bad

**Review by scottinet — APPROVED** [review]

**Review by Aschen — APPROVED** [review]

**Pull request merged**